### PR TITLE
Add monsters walking back to spawn

### PR DESCRIPTION
**Pull request by marmichalski**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Monsters will now walk back to their spawn position when lured away and no targets are around. See the attached videos for more details. When the path gets blocked, monster will just stop.

`spawnwalk.mp4` - monster does not walk back when in allowed range, and walk backs when goes outside of it.
`spawnwalk_block.mp4` - monster cancels walk to spawn when target appears | stop when walking back and path gets blocked.

https://user-images.githubusercontent.com/57528542/131175910-c9ec25c6-89cc-4dda-bee6-df0428018947.mp4

https://user-images.githubusercontent.com/57528542/131175972-af2d6fb3-69c3-4b32-91b1-d8c96ee8cdec.mp4

**Issues addressed:** Part of #922, #3600

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Comment by Bipbop69** [comment]
`home/testeserver/src/monster.cpp: In member function ‘void Monster::onCreatureLeave(Creature*)’:
/home/testeserver/src/monster.cpp:536:55: error: no matching function for call to ‘Monster::getPathTo(Position&, std::vector<Direction>&, int, int, bool, bool)’
     if (getPathTo(masterPos, dirList, 0, 0, true, true)) {
                                                       ^
In file included from /home/testeserver/src/monsters.h:23:0,
                 from /home/testeserver/src/monster.h:24,
                 from /home/testeserver/src/monster.cpp:24:
/home/testeserver/src/creature.h:503:8: note: candidate: bool Creature::getPathTo(const Position&, std::forward_list<Direction>&, const FindPathParams&) const
   bool getPathTo(const Position& targetPos, std::forward_list<Direction>& dirList, const FindPathParams& fpp) const;
        ^~~~~~~~~
/home/testeserver/src/creature.h:503:8: note:   candidate expects 3 arguments, 6 provided
/home/testeserver/src/creature.h:504:8: note: candidate: bool Creature::getPathTo(const Position&, std::forward_list<Direction>&, int32_t, int32_t, bool, bool, int32_t) const
   bool getPathTo(const Position& targetPos, std::forward_list<Direction>& dirList, int32_t minTargetDist, int32_t maxTargetDist, bool fullPathSearch = true, bool clearSight = true, int32_t maxSearchDist = 0) const;
        ^~~~~~~~~
/home/testeserver/src/creature.h:504:8: note:   no known conversion for argument 2 from ‘std::vector<Direction>’ to ‘std::forward_list<Direction>&’
/home/testeserver/src/monster.cpp:537:27: error: no matching function for call to ‘Monster::startAutoWalk(std::vector<Direction>&)’
      startAutoWalk(dirList);
                           ^
In file included from /home/testeserver/src/monsters.h:23:0,
                 from /home/testeserver/src/monster.h:24,
                 from /home/testeserver/src/monster.cpp:24:
/home/testeserver/src/creature.h:289:8: note: candidate: void Creature::startAutoWalk(const std::forward_list<Direction>&)
   void startAutoWalk(const std::forward_list<Direction>& listDir);
        ^~~~~~~~~~~~~
`

File: src/configmanager.cpp [hunk 1]
```diff
@@ -258,8 +259,9 @@ bool ConfigManager::load()
 
 	integer[MAX_PLAYERS] = getGlobalNumber(L, "maxPlayers");
 	integer[PZ_LOCKED] = getGlobalNumber(L, "pzLocked", 60000);
-	integer[DEFAULT_DESPAWNRANGE] = getGlobalNumber(L, "deSpawnRange", 2);
-	integer[DEFAULT_DESPAWNRADIUS] = getGlobalNumber(L, "deSpawnRadius", 50);
+	integer[DEFAULT_DESPAWNRANGE] = Monster::despawnRange = getGlobalNumber(L, "deSpawnRange", 2);
```
Comment (MillhioreBT) [inline src/configmanager.cpp:262]: Here it looks much better ;), now I like it.

I heard that this feature is from Tibia RL, so after all, I will be happy to see this change, I have no more things to say, I hope others can try and give their opinions.

I still have some doubts about what would happen if someone tried to make them walk very long distances? any definite behavior or just monsters walking crazy all over the place?

**Review by EPuncker — APPROVED** [review]

**Comment by EPuncker** [comment]
> `home/testeserver/src/monster.cpp: In member function ‘void Monster::onCreatureLeave(Creature*)’: /home/testeserver/src/monster.cpp:536:55: error: no matching function for call to ‘Monster::getPathTo(Position&, std::vector<Direction>&, int, int, bool, bool)’ if (getPathTo(masterPos, dirList, 0, 0, true, true)) { ^ In file included from /home/testeserver/src/monsters.h:23:0, from /home/testeserver/src/monster.h:24, from /home/testeserver/src/monster.cpp:24: /home/testeserver/src/creature.h:503:8: note: candidate: bool Creature::getPathTo(const Position&, std::forward_list<Direction>&, const FindPathParams&) const bool getPathTo(const Position& targetPos, std::forward_list<Direction>& dirList, const FindPathParams& fpp) const; ^~~~~~~~~ /home/testeserver/src/creature.h:503:8: note: candidate expects 3 arguments, 6 provided /home/testeserver/src/creature.h:504:8: note: candidate: bool Creature::getPathTo(const Position&, std::forward_list<Direction>&, int32_t, int32_t, bool, bool, int32_t) const bool getPathTo(const Position& targetPos, std::forward_list<Direction>& dirList, int32_t minTargetDist, int32_t maxTargetDist, bool fullPathSearch = true, bool clearSight = true, int32_t maxSearchDist = 0) const; ^~~~~~~~~ /home/testeserver/src/creature.h:504:8: note: no known conversion for argument 2 from ‘std::vector<Direction>’ to ‘std::forward_list<Direction>&’ /home/testeserver/src/monster.cpp:537:27: error: no matching function for call to ‘Monster::startAutoWalk(std::vector<Direction>&)’ startAutoWalk(dirList); ^ In file included from /home/testeserver/src/monsters.h:23:0, from /home/testeserver/src/monster.h:24, from /home/testeserver/src/monster.cpp:24: /home/testeserver/src/creature.h:289:8: note: candidate: void Creature::startAutoWalk(const std::forward_list<Direction>&) void startAutoWalk(const std::forward_list<Direction>& listDir); ^~~~~~~~~~~~~ `

the code compiled just fine here, you either did something wrong or you are not using tfs-master but other engine

**Review by DSpeichert — APPROVED** [review]

**Comment by MillhioreBT** [comment]
The PR does not work at all well, only movements of 6 or 7 SQM maximum are allowed due to the getPathTo function So the PR itself is useless and does not work in 99% of cases

Please change the mechanism to the onThink event and have the monster update getPathTo until it reaches its destination
I hope you have a nice day @marmichalski ;)

**Pull request merged**